src/scanpy/preprocessing/_simple.py

```diff
@@ -612,6 +612,53 @@ def normalize_per_cell(  # noqa: PLR0917
     return X if copy else None
 
 
+DT = TypeVar("DT")
+
+
+@numba.njit(cache=True, parallel=True)
+def to_dense(
+    shape: tuple[int, int],
+    indptr: NDArray[np.integer],
+    indices: NDArray[np.integer],
+    data: NDArray[np.number],
```

Comment:
```suggestion
    data: NDArray[DT],
```

sorry, that’s what I meant! “the datatype stays the same” is the point here.